v2.0.0 if you want it?

brianbillman opened the pull request:
Hi, I noticed there was some masked email functionality missing from the CLI (description, delete), so I took a stab at adding them, and ended up doing a bit more than that (I tried to clean things up/refactor a bit to what made more sense *to me*).  It's mostly the same, but not entirely backwards compatible (some command arguments I changed around to convert positional args to named), so I called this v2.0.0.  This was my first go at... go, was a bit of fun, but I tried not to mess with any of the more involved parts.  

There was only one area that was a bit of a hack, for clearing out a domain/description, it sets it to a single space, otherwise it'd never get passed (due to the `omitempty`, but I wanted to keep that in (the space gets stripped when displaying though, so it appears empty to user).  

If you don't want any of the changes, no worries!  I've got an [Alfred](https://www.alfredapp.com) (for Mac) workflow to put up soon as well that uses this new version for quick access to just about all the features.  

Here's the updated help/usage as an overview:
```
Usage of ./maskedemail-cli:
Flags:
  -accountid string
    	fastmail account id (or MASKEDEMAIL_ACCOUNTID env)
  -appname string
    	the appname to identify the creator (or MASKEDEMAIL_APPNAME env) (default: maskedemail-cli)
  -token string
    	the token to authenticate with (or MASKEDEMAIL_TOKEN env)
  -version
    	display the version of maskedemail-cli

Commands:
  maskedemail-cli create [-domain "<domain>"] [-desc "<description>"] [-enabled=true|false (default true)]
  maskedemail-cli list [-show-deleted] [-all-fields]
  maskedemail-cli enable <maskedemail>
  maskedemail-cli disable <maskedemail>
  maskedemail-cli delete <maskedemail>
  maskedemail-cli update -email <maskedemail> [-domain "<domain>"] [-desc "<description>"]
  maskedemail-cli session
```

brianbillman committed:
Add Description field into create command

Also changed List command to show "Description" and "Created At" fields

brianbillman committed:
Expand functionality, refactor

brianbillman committed:
Rename methods/variables "ForDomain" to "Domain"

brianbillman committed:
Clean up success messages slightly

brianbillman committed:
Change to flag params for creation option

Create option can not be used with 0, 1, 2, or 3 parameters, in any order.

Flags:
-domain "<domain>" : the domain the masked email will be used for; optional
-desc "<description>" : the description of the masked email; optional
-enabled=(true|false) : is the newly created masked email enabled; optional, defaults to true

Also shorten "description" to "desc" in command line usage.

brianbillman committed:
Add delete command

brianbillman committed:
Change `-show-deleted` arg to be `list` command specific

Only applicable to this command, make it so

brianbillman committed:
Combine `update` commands into one with named arguments

Also refactor internal methods related to update

brianbillman committed:
Clean up in-command usage and output for help/errors

brianbillman committed:
Update README to reflect new commands and changes

brianbillman committed:
Misc cleanup on commented code

brianbillman committed:
Add version command and capability to makefile

brianbillman committed:
Variable declaration & misc cleanup

brianbillman committed:
Add `all-fields` flag to `list` command

brianbillman committed:
Update Readme

brianbillman committed:
Rearrange order of columns in `list` command

dvcrn commented:
Thanks for the contribution! I'll try to get this reviewed in the next few days

dvcrn commented:
I've tried to review this, but it looks like none of the additional flags you added are working for me. For example: 

```
go run main.go -token xxx -accountid xxx update -email "somemail@domain.com" -domain "foobar"

Usage of update:
  -desc string
        description for the masked email (optional, only updated if argument passed)
  -domain string
        domain for the masked email (optional, only updated if argument passed)
  -email string
        masked email to update (required)
exit status 1
```

brianbillman commented:
Hi, the issue is the`-token xxx` and `-accountid xxx` on the command line is throwing part of it off.  I never tested with `accountid` (it was never needed), just `token` as an env variable.  I'll need to look at it to resolve the issue, it's due to unexpected number of command line args.  If you have it set as an env variable it will work in the meantime.

brianbillman committed:
Fix issue with global flags messing up subsequent flags/args

See: https://stackoverflow.com/questions/67904765/global-flags-and-subcommands

brianbillman committed:
Change `version` from a flag to a command

Makes it cleaner, since it always just outputs the version and that's it

brianbillman committed:
Update README

brianbillman commented:
I have fixed the issue as far as I can tell.  I also changed `-version` from a global flag to a command (`version`), as it made more sense that way.  New usage:

```
Usage of ./maskedemail-cli:
Global Flags:
  -accountid string
    	fastmail account id (or MASKEDEMAIL_ACCOUNTID env)
  -appname string
    	the appname to identify the creator (or MASKEDEMAIL_APPNAME env) (default: maskedemail-cli)
  -token string
    	the token to authenticate with (or MASKEDEMAIL_TOKEN env)

Commands:
  maskedemail-cli create [-domain "<domain>"] [-desc "<description>"] [-enabled=true|false (default true)]
  maskedemail-cli list [-show-deleted] [-all-fields]
  maskedemail-cli enable <maskedemail>
  maskedemail-cli disable <maskedemail>
  maskedemail-cli delete <maskedemail>
  maskedemail-cli update -email <maskedemail> [-domain "<domain>"] [-desc "<description>"]
  maskedemail-cli session
  maskedemail-cli version
```

brianbillman commented:
I should also note that I added into the `make` build versioning and several options around that, to include the current git tag (if one exists) as the version (`vX.Y.Z`), and the commit hash as well.  

Example:
```
% make clean; make version-patch
rm -rf bin/
mkdir bin/ &> /dev/null || true
Building version 2.0.1-b8631a8 of application:
go build -ldflags "-X 'main.buildVersion=2.0.1' -X 'main.buildCommit=b8631a8'" -o bin/maskedemail-cli


% ./bin/maskedemail-cli version 
version: 2.0.1
commit: b8631a8
```

dvcrn commented:
Apologies for the delay and thanks for updating. I started reviewing the PR but needed a bit more time because of the big change

I noticed some smaller things like the version command not actually working because `token` validation happens beforehand, so it just fails if no token is specified. That, and some smaller things within requests, and no gofmt. 

I may just merge this on the weekend and correct the remaining points directly to not have this stall for so long 🙂

dvcrn commented:
I've done a few changes so the code matches better with the other parts, but finally got around to merging this. My changes were mainly: 

- Fix version command requiring token
- Change `update` to have maskedemail positional argument instead of flag, to match with enable/disable
- Applied gofmt to everything
- Refactored UpdateFields to use optional arguments 

Sorry for taking so long and thanks a lot for taking the time to contribute!
